File: lib/cube.ml
type sticker = YELLOW | WHITE | BLUE | RED | GREEN | ORANGE
[@@deriving yojson]

type sticker_row = { fst : sticker; snd : sticker; trd : sticker }
[@@deriving yojson]

type layer = {
  front : sticker_row;
  right : sticker_row;
  back : sticker_row;
  left : sticker_row;
}
[@@deriving yojson]

type face = { fst : sticker_row; snd : sticker_row; trd : sticker_row }
[@@deriving yojson]

type cube = {
  top_face : face;
  top_layer : layer;
  middle_layer : layer;
  bottom_layer : layer;
  bottom_face : face;
}
[@@deriving yojson]

let solved_cube_layer =
  {
    front = { fst = GREEN; snd = GREEN; trd = GREEN };
    right = { fst = RED; snd = RED; trd = RED };
    back = { fst = BLUE; snd = BLUE; trd = BLUE };
    left = { fst = ORANGE; snd = ORANGE; trd = ORANGE };
  }

let solved_cube =
  {
    top_face =
      {
        fst = { fst = WHITE; snd = WHITE; trd = WHITE };
        snd = { fst = WHITE; snd = WHITE; trd = WHITE };
        trd = { fst = WHITE; snd = WHITE; trd = WHITE };
      };
    top_layer = solved_cube_layer;
    middle_layer = solved_cube_layer;
    bottom_layer = solved_cube_layer;
    bottom_face =
      {
        fst = { fst = YELLOW; snd = YELLOW; trd = YELLOW };
        snd = { fst = YELLOW; snd = YELLOW; trd = YELLOW };
        trd = { fst = YELLOW; snd = YELLOW; trd = YELLOW };
      };
  }

let sticker_to_string = function
  | YELLOW -> "Y"
  | WHITE -> "W"
  | BLUE -> "B"
  | RED -> "R"
  | GREEN -> "G"
  | ORANGE -> "O"

let cube_to_string cube =
  Printf.sprintf
    "\n\
    \         %s %s %s\n\
    \         %s %s %s\n\
    \         %s %s %s\n\n\
    \  %s %s %s  %s %s %s  %s %s %s  %s %s %s\n\
    \  %s %s %s  %s %s %s  %s %s %s  %s %s %s\n\
    \  %s %s %s  %s %s %s  %s %s %s  %s %s %s\n\n\
    \         %s %s %s\n\
    \         %s %s %s\n\
    \         %s %s %s\n\
    \  "
    (sticker_to_string cube.top_face.fst.fst)
    (sticker_to_string cube.top_face.fst.snd)
    (sticker_to_string cube.top_face.fst.trd)
    (sticker_to_string cube.top_face.snd.fst)
    (sticker_to_string cube.top_face.snd.snd)
    (sticker_to_string cube.top_face.snd.trd)
    (sticker_to_string cube.top_face.trd.fst)
    (sticker_to_string cube.top_face.trd.snd)
    (sticker_to_string cube.top_face.trd.trd)
    (sticker_to_string cube.top_layer.left.fst)
    (sticker_to_string cube.top_layer.left.snd)
    (sticker_to_string cube.top_layer.left.trd)
    (sticker_to_string cube.top_layer.front.fst)
    (sticker_to_string cube.top_layer.front.snd)
    (sticker_to_string cube.top_layer.front.trd)
    (sticker_to_string cube.top_layer.right.fst)
    (sticker_to_string cube.top_layer.right.snd)
    (sticker_to_string cube.top_layer.right.trd)
    (sticker_to_string cube.top_layer.back.fst)
    (sticker_to_string cube.top_layer.back.snd)
    (sticker_to_string cube.top_layer.back.trd)
    (sticker_to_string cube.middle_layer.left.fst)
    (sticker_to_string cube.middle_layer.left.snd)
    (sticker_to_string cube.middle_layer.left.trd)
    (sticker_to_string cube.middle_layer.front.fst)
    (sticker_to_string cube.middle_layer.front.snd)
    (sticker_to_string cube.middle_layer.front.trd)
    (sticker_to_string cube.middle_layer.right.fst)
    (sticker_to_string cube.middle_layer.right.snd)
    (sticker_to_string cube.middle_layer.right.trd)
    (sticker_to_string cube.middle_layer.back.fst)
    (sticker_to_string cube.middle_layer.back.snd)
    (sticker_to_string cube.middle_layer.back.trd)
    (sticker_to_string cube.bottom_layer.left.fst)
    (sticker_to_string cube.bottom_layer.left.snd)
    (sticker_to_string cube.bottom_layer.left.trd)
    (sticker_to_string cube.bottom_layer.front.fst)
    (sticker_to_string cube.bottom_layer.front.snd)
    (sticker_to_string cube.bottom_layer.front.trd)
    (sticker_to_string cube.bottom_layer.right.fst)
    (sticker_to_string cube.bottom_layer.right.snd)
    (sticker_to_string cube.bottom_layer.right.trd)
    (sticker_to_string cube.bottom_layer.back.fst)
    (sticker_to_string cube.bottom_layer.back.snd)
    (sticker_to_string cube.bottom_layer.back.trd)
    (sticker_to_string cube.bottom_face.fst.fst)
    (sticker_to_string cube.bottom_face.fst.snd)
    (sticker_to_string cube.bottom_face.fst.trd)
    (sticker_to_string cube.bottom_face.snd.fst)
    (sticker_to_string cube.bottom_face.snd.snd)
    (sticker_to_string cube.bottom_face.snd.trd)
    (sticker_to_string cube.bottom_face.trd.fst)
    (sticker_to_string cube.bottom_face.trd.snd)
    (sticker_to_string cube.bottom_face.trd.trd)

let cross_top_face_is_solved cube =
  let is_cross =
    cube.top_face.snd.snd == cube.top_face.fst.snd
    && cube.top_face.snd.snd == cube.top_face.snd.fst
    && cube.top_face.snd.snd == cube.top_face.snd.trd
    && cube.top_face.snd.snd == cube.top_face.trd.snd
  in
  let is_matching_edges =
    cube.top_layer.front.snd == cube.middle_layer.front.snd
    && cube.top_layer.right.snd == cube.middle_layer.right.snd
    && cube.top_layer.left.snd == cube.middle_layer.left.snd
    && cube.top_layer.back.snd == cube.middle_layer.back.snd
  in
  is_cross && is_matching_edges

let cross_bottom_face_is_solved cube =
  let is_cross =
    cube.bottom_face.snd.snd == cube.bottom_face.fst.snd
    && cube.bottom_face.snd.snd == cube.bottom_face.snd.fst
    && cube.bottom_face.snd.snd == cube.bottom_face.snd.trd
    && cube.bottom_face.snd.snd == cube.bottom_face.trd.snd
  in
  let is_matching_edges =
    cube.bottom_layer.front.snd == cube.middle_layer.front.snd
    && cube.bottom_layer.right.snd == cube.middle_layer.right.snd
    && cube.bottom_layer.left.snd == cube.middle_layer.left.snd
    && cube.bottom_layer.back.snd == cube.middle_layer.back.snd
  in
  is_cross && is_matching_edges

let corners_top_layer_are_solved cube =
  let corners_inserted =
    cube.top_face.snd.snd == cube.top_face.fst.fst
    && cube.top_face.snd.snd == cube.top_face.fst.trd
    && cube.top_face.snd.snd == cube.top_face.trd.fst
    && cube.top_face.snd.snd == cube.top_face.trd.trd
  in
  let is_matching_edges =
    cube.top_layer.front.fst == cube.middle_layer.front.snd
    && cube.top_layer.front.trd == cube.middle_layer.front.snd
    && cube.top_layer.right.fst == cube.middle_layer.right.snd
    && cube.top_layer.right.trd == cube.middle_layer.right.snd
    && cube.top_layer.left.fst == cube.middle_layer.left.snd
    && cube.top_layer.left.trd == cube.middle_layer.left.snd
    && cube.top_layer.back.fst == cube.middle_layer.back.snd
    && cube.top_layer.back.trd == cube.middle_layer.back.snd
  in
  corners_inserted && is_matching_edges

let corners_bottom_layer_are_solved cube =
  let corners_inserted =
    cube.bottom_face.snd.snd == cube.bottom_face.fst.fst
    && cube.bottom_face.snd.snd == cube.bottom_face.fst.trd
    && cube.bottom_face.snd.snd == cube.bottom_face.trd.fst
    && cube.bottom_face.snd.snd == cube.bottom_face.trd.trd
  in
  let is_matching_edges =
    cube.bottom_layer.front.fst == cube.middle_layer.front.snd
    && cube.bottom_layer.front.trd == cube.middle_layer.front.snd
    && cube.bottom_layer.right.fst == cube.middle_layer.right.snd
    && cube.bottom_layer.right.trd == cube.middle_layer.right.snd
    && cube.bottom_layer.left.fst == cube.middle_layer.left.snd
    && cube.bottom_layer.left.trd == cube.middle_layer.left.snd
    && cube.bottom_layer.back.fst == cube.middle_layer.back.snd
    && cube.bottom_layer.back.trd == cube.middle_layer.back.snd
  in
  corners_inserted && is_matching_edges

let edges_second_layer_are_solved cube =
  cube.middle_layer.front.fst == cube.middle_layer.front.snd
  && cube.middle_layer.front.trd == cube.middle_layer.front.snd
  && cube.middle_layer.right.fst == cube.middle_layer.right.snd
  && cube.middle_layer.right.trd == cube.middle_layer.right.snd
  && cube.middle_layer.left.fst == cube.middle_layer.left.snd
  && cube.middle_layer.left.trd == cube.middle_layer.left.snd
  && cube.middle_layer.back.fst == cube.middle_layer.back.snd
  && cube.middle_layer.back.trd == cube.middle_layer.back.snd

File: frontend/src/lib/cube.ts
export enum Sticker {
  YELLOW,
  WHITE,
  BLUE,
  RED,
  GREEN,
  ORANGE,
}
type StickerRow = { fst : Sticker; snd : Sticker; trd : Sticker }
type Layer = {
  front: StickerRow;
  right: StickerRow;
  back: StickerRow;
  left: StickerRow;
}
type Face = { fst : StickerRow; snd : StickerRow; trd : StickerRow }
export type Cube = {
  top_face: Face;
  top_layer: Layer;
  middle_layer: Layer;
  bottom_layer: Layer;
  bottom_face: Face;
};

function stickerFromJson(json: any): Sticker {
  const sticker_to_color_map = new Map<string, Sticker>([
    ["BLUE", Sticker.BLUE],
    ["GREEN", Sticker.GREEN],
    ["RED", Sticker.RED],
    ["YELLOW", Sticker.YELLOW],
    ["WHITE", Sticker.WHITE],
    ["ORANGE", Sticker.ORANGE],
  ]);
  const sticker = sticker_to_color_map.get(json[0]);
  if (sticker == undefined) {
    throw Error("Unable to parse json as Cube")
  }
  return sticker;
}

function stickerToJson(sticker: Sticker) {
  const sticker_to_color_map = new Map<Sticker, string[]>([
    [Sticker.BLUE, ["BLUE"]],
    [Sticker.GREEN, ["GREEN"]],
    [Sticker.RED, ["RED"]],
    [Sticker.YELLOW, ["YELLOW"]],
    [Sticker.WHITE, ["WHITE"]],
    [Sticker.ORANGE, ["ORANGE"]],
  ]);
  return sticker_to_color_map.get(sticker);
}

function rowFromJson(json: any): StickerRow {
  return {
    fst: stickerFromJson(json.fst),
    snd: stickerFromJson(json.snd),
    trd: stickerFromJson(json.trd),
  };
}

function rowToJson(row: StickerRow) {
  return {
    fst: stickerToJson(row.fst),
    snd: stickerToJson(row.snd),
    trd: stickerToJson(row.trd),
  };
}

function layerFromJson(json: any): Layer {
  return {
    front: rowFromJson(json.front),
    right: rowFromJson(json.right),
    back: rowFromJson(json.back),
    left: rowFromJson(json.left),
  };
}

function layerToJson(layer: Layer) {
  return {
    front: rowToJson(layer.front),
    right: rowToJson(layer.right),
    back: rowToJson(layer.back),
    left: rowToJson(layer.left),
  };
}

function faceFromJson(json: any): Face {
  return {
    fst: rowFromJson(json.fst),
    snd: rowFromJson(json.snd),
    trd: rowFromJson(json.trd),
  };
}

function faceToJson(face: Face) {
  return {
    fst: rowToJson(face.fst),
    snd: rowToJson(face.snd),
    trd: rowToJson(face.trd),
  };
}

export function cubeFromJson(json: any): Cube {
  return {
    top_face: faceFromJson(json.top_face), 
    top_layer: layerFromJson(json.top_layer), 
    middle_layer: layerFromJson(json.middle_layer), 
    bottom_layer: layerFromJson(json.bottom_layer), 
    bottom_face: faceFromJson(json.bottom_face), 
  }
}

export function cubeToJson(cube: Cube) {
  return {
    top_face: faceToJson(cube.top_face), 
    top_layer: layerToJson(cube.top_layer), 
    middle_layer: layerToJson(cube.middle_layer), 
    bottom_layer: layerToJson(cube.bottom_layer), 
    bottom_face: faceToJson(cube.bottom_face), 
  }
}
